### ISSUE TITLE
用paddle神经网络做二分类时，拿训练集一些样本做预测，probs输出很奇怪，预测全为0，而我用sklearn 逻辑回归输出预测是正常的

### ISSUE THREAD

**Issue by MEYOUTH** [PaddlePaddle/Paddle#8112]
probs = paddle.infer(output_layer=predict, parameters=parametersTotal, input=train_matrix[500:530,:])
probs如下：
[[  9.91644561e-01   8.35537445e-03]
 [  9.66482282e-01   3.35177518e-02]
 [  9.90382254e-01   9.61771514e-03]
 [  9.98637259e-01   1.36269839e-03]
 [  9.90097940e-01   9.90204606e-03]
 [  9.99152303e-01   8.47698189e-04]
 [  9.99365389e-01   6.34588883e-04]
 [  9.98789489e-01   1.21054577e-03]
 [  9.70943928e-01   2.90561188e-02]
 [  9.85457122e-01   1.45428479e-02]
 [  9.83285725e-01   1.67142693e-02]
 [  9.99981999e-01   1.80263087e-05]
 [  9.99694467e-01   3.05603491e-04]
 [  9.97109711e-01   2.89028184e-03]
 [  9.99974132e-01   2.58297641e-05]
 [  9.98865128e-01   1.13483786e-03]
 [  9.98078585e-01   1.92143000e-03]
 [  9.99985933e-01   1.40257634e-05]
 [  9.97092962e-01   2.90708337e-03]
 [  9.99721348e-01   2.78617663e-04]
 [  9.96503115e-01   3.49691231e-03]
 [  9.99973893e-01   2.60580273e-05]
 [  9.97609973e-01   2.39002821e-03]
 [  9.93797839e-01   6.20210869e-03]
 [  9.96902525e-01   3.09744780e-03]
 [  9.96561110e-01   3.43887508e-03]
 [  9.99003947e-01   9.96033312e-04]
 [  9.99252617e-01   7.47355691e-04]
 [  9.99153376e-01   8.46567214e-04]
 [  9.94594157e-01   5.40590566e-03]]

**Comment by peterzhang2029** [comment]
你的数据类型是什么？ 贴一下你的配置吧

**Comment by MEYOUTH** [comment]
x = paddle.layer.data(name='x', type=paddle.data_type.dense_vector(100))
#第一个全连接层，激活函数为ReLU
hidden1 = paddle.layer.fc(input=x, size=32, act=paddle.activation.Relu())
# 第二个全连接层，激活函数为ReLU
hidden2 = paddle.layer.fc(input=hidden1,
                              size=16,
                              act=paddle.activation.Relu())
predict = paddle.layer.fc(input=hidden2,
                              size=2,
                              act=paddle.activation.Softmax())
y = paddle.layer.data(name='y', type=paddle.data_type.integer_value(2))
cost = paddle.layer.classification_cost(input=predict, label=y)
                              act=paddle.activation.Softmax())
y = paddle.layer.data(name='y', type=paddle.data_type.integer_value(2))
cost = paddle.layer.classification_cost(input=predict, label=y)

**Comment by MEYOUTH** [comment]
我是照着官网数字识别的例子来，用了最简单的softmax也不行，全预测为一样的，实在找不到错误

**Comment by peterzhang2029** [comment]
用的什么数据， 先检查下你的数据是否有问题

**Comment by MEYOUTH** [comment]
都是浮点型，大部分是0到1之间，稠密型的

**Comment by peterzhang2029** [comment]
可以通过 https://github.com/PaddlePaddle/Paddle/blob/29603cf39a8d61475dee013b23b6aa589d75887f/doc/faq/local/index_cn.rst#layeroutput 这里先看看每一层的输出或者参数 是否有异常

**Comment by yeyupiaoling** [comment]
@MEYOUTH 
训练输出的Test 是否正常？在预测是有没有对数据集做了其他的处理

**Comment by MEYOUTH** [comment]
训练集数据没有做其他处理，数据大部分都在0,1之间的浮点数

**Comment by MEYOUTH** [comment]
用测试集预测也不正常

**Comment by MEYOUTH** [comment]
想问下二分类时，paddle对训练集四个样本y是用[1,0,0,1]这样表示，还是像tensorflow中[[1,0] [0,1],[0,1],[1,0]] @peterzhang2029

**Comment by peterzhang2029** [comment]
@MEYOUTH 是前者， 只需要一个标记

**Comment by MEYOUTH** [comment]
如何做标记呢，我一直有个疑问，比如三分类时，y的值随便给吗，paddle.infer得到probs的每一列对应是那个类呢 @peterzhang2029

**Comment by peterzhang2029** [comment]
比如三个样本分别属于第1、2、3类，可以组织为[0,1,2], infer出的结果也是对应类别的概率。

**Comment by MEYOUTH** [comment]
我在本地试过，比如二分类把y=【0,0,1】改成【1,1,0】，但是输出的probs还是一样，输出的第一列和第二列并没有颠倒

**Comment by peterzhang2029** [comment]
@MEYOUTH 你是用训练好的模型去进行测试集的类别预测么，你的训练数据很足够大么？ 或者你可以发你的模型配置看一下。

**Comment by MEYOUTH** [comment]
# coding=UTF-8
import numpy as np
# import matplotlib.pyplot as plt
import cPickle
import sys
import json
import paddle.v2 as paddle
Train_matrix = np.array([[0,3,1],[2,2,2],[6,4,5]])
mt1=np.array([[0,3,1]])
mt2=np.array([[2,2,2]])
mt3=np.array([[6,4,5]])
mt4=np.array([[10,2,5]])
mts=[[mt1],[mt2],[mt3],[mt4]]

# we=np.array([[0,3,1]])
# we1=np.array([[2,2,2]])
# wes=[we,we1,]
label_matrix= [0,0,1]
# label_matrix= np.array([[0,1], [0,1], [1,0]])
paddle.init(use_gpu=False, trainer_count=1)
x = paddle.layer.data(name='x', type=paddle.data_type.dense_vector(3))
predict = paddle.layer.fc(input=x,
                              size=2,
                              act=paddle.activation.Softmax())
y = paddle.layer.data(name='y', type=paddle.data_type.integer_value(2))
cost = paddle.layer.classification_cost(input=predict, label=y)

parameters = paddle.parameters.create(cost)
optimizer = paddle.optimizer.Momentum(
    learning_rate=0.1 / 128.0,
    momentum=0.9,
    regularization=paddle.optimizer.L2Regularization(rate=0.0005 * 128))
trainer = paddle.trainer.SGD(cost=cost,
                             parameters=parameters,
                             update_equation=optimizer)
# define training dataset reader
def train_reader():
    train_x = np.array([[0,3,1],[2,2,2],[6,4,5]])
    train_y = label_matrix
    def reader():
        for i in xrange(train_x.shape[0]):#因为train_y为list型，所以不能用shape，改为train_x.shape[0]
            yield train_x[i], train_y[i]
    return reader

# define feeding map
feeding = {'x': 0, 'y': 1}
step=0
train_title = "Train cost"
test_title = "Test cost"
# cost_ploter = Ploter(train_title)
estep=0
# event_handler to print training info
def event_handler(event):
    global step
    global estep
    if isinstance(event, paddle.event.EndIteration):
        if event.batch_id % 1 == 0:
            print "Pass %d, Batch %d, Cost %f" % (
                event.pass_id, event.batch_id, event.cost)
    if isinstance(event, paddle.event.EndPass):
        estep += 1
        if (estep == 9):
            with open('scnnparams_pass_%d.tar' % 128, 'w') as f:
                parameters.to_tar(f)
trainer.train(
    reader=paddle.batch(train_reader(), batch_size=100),
    feeding=feeding,
    event_handler=event_handler,
    num_passes=10)
print Train_matrix.shape
probs = paddle.infer(output_layer=predict, parameters=parameters, input=mts)
print probs

**Comment by MEYOUTH** [comment]
配置是这样的
paddle.init(use_gpu=False, trainer_count=1)
x = paddle.layer.data(name='x', type=paddle.data_type.dense_vector(3))
predict = paddle.layer.fc(input=x,
                              size=2,
                              act=paddle.activation.Softmax())
y = paddle.layer.data(name='y', type=paddle.data_type.integer_value(2))
cost = paddle.layer.classification_cost(input=predict, label=y)

parameters = paddle.parameters.create(cost)
optimizer = paddle.optimizer.Momentum(
    learning_rate=0.1 / 128.0,
    momentum=0.9,
    regularization=paddle.optimizer.L2Regularization(rate=0.0005 * 128))
trainer = paddle.trainer.SGD(cost=cost,
                             parameters=parameters,
                             update_equation=optimizer)
@peterzhang2029

**Comment by peterzhang2029** [comment]
我用的你给的程序试了一下 问题出现在训练数据太少了 模型根本就没办法从训练数据中学习到东西。 可以将数据增多一些试试。

```
# coding=UTF-8
import numpy as np
# import matplotlib.pyplot as plt
import cPickle
import sys
import json
import paddle.v2 as paddle
np.random.seed(1)
Train_matrix = np.array([[0,3,1],[2,2,2],[6,4,5]])#训练用的
Train_matrix = np.random.rand(1000, 3)#训练用的
mt1=np.array([[0,3,1]])
mt2=np.array([[2,2,2]])
mt3=np.array([[6,4,5]])
mt4=np.array([[10,2,5]])
mts=[[mt1],[mt2],[mt3],[mt4]]#测试用的

#label_matrix = [1,1,0]#这是y,改这里了,替换成【1,1,0】，probs还是一样
np.random.seed(1)
label_matrix = np.random.randint(0,1,size=(1000, 1))
# label_matrix= np.array([[0,1], [0,1], [1,0]])
paddle.init(use_gpu=False, trainer_count=1)
x = paddle.layer.data(name='x', type=paddle.data_type.dense_vector(3))
predict = paddle.layer.fc(input=x,
                              size=2,
                              act=paddle.activation.Softmax())
y = paddle.layer.data(name='y', type=paddle.data_type.integer_value(2))
cost = paddle.layer.classification_cost(input=predict, label=y)

parameters = paddle.parameters.create(cost)
optimizer = paddle.optimizer.Momentum(
    learning_rate=0.1 / 128.0,
    momentum=0.9,
    regularization=paddle.optimizer.L2Regularization(rate=0.0005 * 128))
trainer = paddle.trainer.SGD(cost=cost,
                             parameters=parameters,
                             update_equation=optimizer)
# define training dataset reader
def train_reader():
    train_x = Train_matrix #np.array([[0,3,1],[2,2,2],[6,4,5]])
    train_y = label_matrix
    def reader():
        for i in xrange(train_x.shape[0]):#
            # inverse the label
            if train_y[i][0] == 1:
                label = 0
            else:
                label = 1
            label = train_y[i][0]
            yield train_x[i], label
            #print train_x[i], train_y[i]
    return reader

# define feeding map
feeding = {'x': 0, 'y': 1}
step=0
train_title = "Train cost"
test_title = "Test cost"
# cost_ploter = Ploter(train_title)
estep=0
# event_handler to print training info
def event_handler(event):
    global step
    global estep
    if isinstance(event, paddle.event.EndIteration):
        if event.batch_id % 1 == 0:
            print "Pass %d, Batch %d, Cost %f" % (
                event.pass_id, event.batch_id, event.cost)
    if isinstance(event, paddle.event.EndPass):
        estep += 1
        if (estep == 9):
            with open('scnnparams_pass_%d.tar' % 128, 'w') as f:
                parameters.to_tar(f)
trainer.train(
    reader=paddle.batch(train_reader(), batch_size=10),
    feeding=feeding,
    event_handler=event_handler,
    num_passes=10)

probs = paddle.infer(output_layer=predict, parameters=parameters, input=mts)
print probs
```

**Comment by peterzhang2029** [comment]
@MEYOUTH 对的

**Comment by peterzhang2029** [comment]
Closing due to low activity. Feel free to reopen it.
